File: app/modules/request_logs/schemas.py
from __future__ import annotations

from datetime import datetime

from pydantic import Field

from app.modules.shared.schemas import DashboardModel


class RequestLogEntry(DashboardModel):
    requested_at: datetime
    account_id: str | None = None
    api_key_name: str | None = None
    request_id: str
    model: str
    transport: str | None = None
    service_tier: str | None = None
    status: str
    error_code: str | None = None
    error_message: str | None = None
    tokens: int | None = None
    cached_input_tokens: int | None = None
    reasoning_effort: str | None = None
    cost_usd: float | None = None
    burn_rate_5h_plus_accounts: float | None = Field(default=None, serialization_alias="burnRate5hPlusAccounts")
    burn_rate_7d_plus_accounts: float | None = Field(default=None, serialization_alias="burnRate7dPlusAccounts")
    latency_ms: int | None = None


class RequestLogsResponse(DashboardModel):
    requests: list[RequestLogEntry] = Field(default_factory=list)
    total: int
    has_more: bool


class RequestLogModelOption(DashboardModel):
    model: str
    reasoning_effort: str | None = None


class RequestLogFilterOptionsResponse(DashboardModel):
    account_ids: list[str] = Field(default_factory=list)
    model_options: list[RequestLogModelOption] = Field(default_factory=list)
    statuses: list[str] = Field(default_factory=list)

File: frontend/src/features/key-dashboard/schemas.ts
import { z } from "zod";

import { RequestLogSchema, type RequestLog } from "@/features/dashboard/schemas";

export const KeyUsageSchema = z.object({
  request_count: z.number().int().nonnegative(),
  total_tokens: z.number().int().nonnegative(),
  cached_input_tokens: z.number().int().nonnegative(),
  total_cost_usd: z.number().nonnegative(),
  limits: z.array(z.unknown()),
  upstream_limits: z.array(z.unknown()).optional(),
  account_pool_usage: z.unknown().nullable().optional(),
});

const KeyDashboardCostBreakdownSchema = z.strictObject({
  inputUsd: z.number().nullable(),
  cachedInputUsd: z.number().nullable(),
  outputUsd: z.number().nullable(),
  totalUsd: z.number().nullable(),
});

export const KeyDashboardRequestLogSchema = z.strictObject({
  requestedAt: z.iso.datetime({ offset: true }),
  requestId: z.string(),
  requestKind: z.enum(["normal", "warmup", "limit_warmup", "prewarm", "compaction", "realtime_live"]),
  model: z.string(),
  transport: z.string().nullable(),
  upstreamTransport: z.string().nullable(),
  serviceTier: z.string().nullable(),
  requestedServiceTier: z.string().nullable(),
  actualServiceTier: z.string().nullable(),
  reasoningEffort: z.string().nullable(),
  status: z.string(),
  errorCode: z.string().nullable(),
  tokens: z.number().int().nullable(),
  inputTokens: z.number().int().nullable(),
  outputTokens: z.number().int().nullable(),
  outputTokensRaw: z.number().int().nullable(),
  reasoningTokens: z.number().int().nullable(),
  cachedInputTokens: z.number().int().nullable(),
  costUsd: z.number().nullable(),
  costBreakdown: KeyDashboardCostBreakdownSchema,
  latencyMs: z.number().int().nullable(),
  latencyFirstTokenMs: z.number().int().nullable(),
  latencyQueueMs: z.number().int().nullable(),
});

export const KeyDashboardRequestLogsResponseSchema = z.strictObject({
  requests: z.array(KeyDashboardRequestLogSchema),
  total: z.number().int().nonnegative(),
  hasMore: z.boolean(),
});

export type KeyUsage = z.infer<typeof KeyUsageSchema>;
export type KeyDashboardRequestLogsResponse = z.infer<typeof KeyDashboardRequestLogsResponseSchema>;

export function toDashboardRequestLog(
  request: z.infer<typeof KeyDashboardRequestLogSchema>,
): RequestLog {
  return RequestLogSchema.parse({
    ...request,
    accountId: null,
    planType: null,
    apiKeyId: null,
    apiKeyName: null,
    archiveRequestId: null,
    connectionRequestKind: null,
    source: null,
    modelSourceId: null,
    modelSourceKind: null,
    upstreamProxyRouteMode: null,
    upstreamProxyPoolId: null,
    upstreamProxyEndpointId: null,
    upstreamProxyFallbackUsed: null,
    upstreamProxyFailClosedReason: null,
    useragent: null,
    useragentGroup: null,
    clientIp: null,
    conversationId: null,
    errorMessage: null,
    failurePhase: null,
    failureDetail: null,
    failureExceptionType: null,
    upstreamStatusCode: null,
    upstreamErrorCode: null,
    bridgeStage: null,
  });
}
